spring-boot-project/spring-boot/src/test/java/org/springframework/boot/context/properties/ConfigurationPropertiesTests.java

```diff
@@ -781,13 +785,16 @@ void loadWhenBindingToConstructorParametersWithCustomDataUnitShouldBind() {
 		source.put("test.duration", "12");
 		source.put("test.size", "13");
 		source.put("test.period", "14");
+		source.put("test.formattedDuration", "15d");
+		source.put("test.formattedPeriod", "16y");
 		sources.addLast(new MapPropertySource("test", source));
 		load(ConstructorParameterWithUnitConfiguration.class);
 		ConstructorParameterWithUnitProperties bean = this.context
 				.getBean(ConstructorParameterWithUnitProperties.class);
 		assertThat(bean.getDuration()).isEqualTo(Duration.ofDays(12));
 		assertThat(bean.getSize()).isEqualTo(DataSize.ofMegabytes(13));
-		assertThat(bean.getPeriod()).isEqualTo(Period.ofYears(14));
+		assertThat(bean.getFormattedDuration()).isEqualTo(Duration.ofDays(15));
+		assertThat(bean.getFormattedPeriod()).isEqualTo(Period.ofYears(16));
```

Comment:
These assertions succeed without `@DurationFormat` and `@PeriodFormat` on `ConstructorParameterWithUnitProperties`. I think that's because the converter successfully infers the style to use from the value that's being converted. To verify that the format annotations are working as expected, I think it would be necessary to trigger a failure by trying, for example, to bind an ISO-8601 value to a property that requires the simple style.

Reply:
I extended/reworked the testing a bit. Let me know if this is what you had in mind, @wilkinsona

Reply:
Spot on. Thanks.